x86: add support for Meraki MX100

riptidewave93 opened the pull request:
This PR adds support for the Meraki MX100, as well as a new package named nu801 to support the status LED on the device.

CC @chunkeey

riptidewave93 committed:
gpio-cdev: add nu801 userspace driver

This adds a userspace interpretation of the nu801 driver used by Meraki
hardware. Previously this was a driver that was added per target, but as
multiple targets now have this driver, we should move to something that
can be shared by all targets since no driver exists upstream.

Co-developed-by: Christian Lamparter <chunkeey@gmail.com>
Signed-off-by: Chris Blake <chrisrblake93@gmail.com>
Signed-off-by: Christian Lamparter <chunkeey@gmail.com>

adschm commented on target/linux/x86/patches-5.10/102-platform-x86-add-meraki-mx100-platform-driver.patch:
```diff
@@ -0,0 +1,300 @@
++		.default_trigger = "default-on",
++	},
++	{
++		.name = "mx100:green:lan2",
```
Consider removing the device prefix so that you just have "green:lan2", "green:internet" etc.

adschm replied:
Or is this already in kernel and represents a backport? Then we probably won't be able to fix it anymore, but then the patch name should contain the upstream version.

riptidewave93 replied:
@adschm both patches are upstream in 5.15. I'll get the patch names updated.

riptidewave93 committed:
x86: add support for Meraki MX100

This commit will add support for the Meraki MX100 in OpenWRT.

Specs:
* CPU: Intel Xeon E3-1200 Series 1.5GHz 2C/4T
* Memory: 4GB DDR3 1600 ECC
* Storage: 1GB USB NAND, 1TB SATA HDD
* Wireless: None
* Wired: 10x 1Gb RJ45, 2x 1Gb SFP

UART:
The UART header is named CONN11 and is found in the
center of the mainboard. The pinout from Pin 1 (marked
with a black triangle) to pin 4 is below:
Pin 1: VCC
Pin 2: TX
Pin 3: RX
Pin 4: GND
Note that VCC is not required for UART on this device.

Booting:
1. Flash/burn one of the images from this repo to a
flash drive.
2. Take the top off the MX100, and unplug the SATA
cable from the HDD.
3. Hook up UART to the MX100, plug in the USB drive,
and then power up the device.
4. At the BIOS prompt, quickly press F7 and then
scroll to the Save & Exit tab.
5. Scroll down to Boot Override, and select the
UEFI entry for your jumpdrive.

Note: UEFI booting will fail if the SATA cable for
the HDD is plugged in.
The issue is explained under the Flashing instructions.

Flashing:
1. Ensure the MX100 is powered down, and not plugged
into power.
2. Take the top off the MX100, and unplug the SATA
cable from the HDD.
3. Using the Mini USB female port found by the SATA
port on the motherboard,
flash one of the images to the system. Example:
`dd if=image of=/dev/sdb conv=fdatasync` where sdb
is the USB device for the MX100's NAND.
4. Unplug the Mini USB, hook up UART to the MX100,
and then power up the device.
5. At the BIOS prompt, quickly press F7 and then
scroll to the Boot tab.
6. Change the boot order and set UEFI: USB DISK 2.0
as first, and USB DISK 2.0 as second.
Disable the other boot options.
7. Go to Save & Exit, and then select Save Changes and
Reset

Note that OpenWRT will fail to boot in UEFI mode when
the SATA hard drive is plugged in. To fix this, boot
with the SATA disk unplugged and then run the following
command:
`sed -i "s|hd0,gpt1|hd1,gpt1|g" boot/grub/grub.cfg`
Once the above is ran, OpenWRT will boot when the HDD
is plugged into SATA. The reason this happens is the
UEFI implementation for the MX100 will always set
anything on SATA to HD0 instead of the onboard USB
storage, so we have to accomidate it since OpenWRT's
GRUB does not support detecting a boot disk via UUID.

Signed-off-by: Chris Blake <chrisrblake93@gmail.com>

hauke commented on package/system/gpio-cdev/nu801/Makefile:
```diff
@@ -0,0 +1,41 @@
+  SUBMENU:=Userspace GPIO Drivers
+  DEPENDS:=@!LINUX_5_4 +kmod-leds-uleds
+  KCONFIG:= \
+    CONFIG_GPIO_CDEV=y
```
I do not like when normal packages activate a Kconfig option. I am not even sure if this works correctly,

riptidewave93 replied:
Would you prefer the generic kernel 5.10 kernel config enables this then? I'm open to changing how this is done, but I want to make sure it's fixed correctly.

chunkeey replied:
for what's it worth: The KCONFIG:= variable is declared in include/package-defaults.mk's ````Package/Default```` section.  which means that every package (doesn't matter if kernel or not) can set it.

Now, about that "not sure if it works". I think It does. (with the caveat that it works in the same way as any other kmod-package. (i.e. for it to work, it has to be included during build as either CONFIG_PACKAGE_NU801=m or =y), so the
requested `CONFIG_GPIO_CDEV` is set. This is done by:

https://github.com/openwrt/openwrt/blob/master/include/kernel-defaults.mk#L112

> [...]package-metadata.pl kconfig $(TMP_DIR)/.packageinfo $(TOPDIR)/.config $(KERNEL_PATCHVER) > $(LINUX_DIR)/.config.override

the package-metadata.pl script which generates the ````.config.override````file which gets passed to the final kernel
(sync?)config step takes that dependency from the same $(TMP_DIR)/.packageinfo as with every other package.

hauke commented:
The nu801 package also gets build on lantiq/xrx200 with kernel 5.4 and then fails, even when It is not selected in the config.

riptidewave93 commented:
> The nu801 package also gets build on lantiq/xrx200 with kernel 5.4 and then fails, even when It is not selected in the config.

I have a hard time understanding how this is possible, but I'll dig into it and see if I can find the cause.

riptidewave93 commented:
@hauke I have been unable to replicate the issue you saw with lantiq/xrx200. Im curious if you have any more information on how you came to find that issue?

chunkeey commented:
Merged. Thanks!

chunkeey commented:
Preliminary revert of gpio-cdev:

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=196a5f54a3dba5507783ed728dc24ae2d3f71780
